sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Snippets/Sample07_MockEventProcessorHandlersTests.cs

```diff
@@ -0,0 +1,244 @@
+    public class Sample07_MockEventProcessorHandlersTests
+    {
+        /// <summary>
+        ///   Performs basic smoke test validation of the contained snippet.
+        /// </summary>
+        ///
+        [Test]
+        public async Task ProcessEventHandlerCheckpoint()
+        {
+            #region Snippet:EventHubs_Processor_Sample07_ProcessEventHandlerCheckpoint
+
+            // Application handler checkpoints after processing defined number of events.
+
+            const int EventsBeforeCheckpoint = 2;
+            var partitionEventCount = new ConcurrentDictionary<string, int>();
+            bool checkpointUpdated = false;
+
+            async Task processEventHandler(ProcessEventArgs args)
+            {
+                try
+                {
+                    checkpointUpdated = false;
+
+                    await Application.ProcessEventAsync(args);
+
+                    string partition = args.Partition.PartitionId;
+
+                    int eventsSinceLastCheckpoint = partitionEventCount.AddOrUpdate(
+                        key: partition,
+                        addValue: 1,
+                        updateValueFactory: (_, currentCount) => currentCount + 1);
+
+                    if (eventsSinceLastCheckpoint >= EventsBeforeCheckpoint)
+                    {
+                        await args.UpdateCheckpointAsync();
+                        partitionEventCount[partition] = 0;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Application.HandleProcessingException(args, ex);
+                }
+            }
+
+            // Initialize the args and simulate checkpoint update method.
+
+            var partitionId = "0";
+            var eventBody = new BinaryData("This is a sample event body");
+            var eventData = EventHubsModelFactory.EventData(eventBody);
+            var partitionContext = EventHubsModelFactory.PartitionContext(partitionId);
+            var eventArgs = new ProcessEventArgs(partitionContext, eventData, _ => { checkpointUpdated = true; return Task.CompletedTask; });
+
+            for (int i = 1; i <= EventsBeforeCheckpoint; i++)
+            {
+                // Execute the handler and validate that the event handling has been checkpointed.
+
+                await processEventHandler(eventArgs);
+
+                if (i % EventsBeforeCheckpoint == 0)
+                    Assert.IsTrue(checkpointUpdated);
+                else
+                    Assert.IsFalse(checkpointUpdated);
+
+                Assert.That(partitionEventCount[partitionId], Is.EqualTo(i % EventsBeforeCheckpoint));
+            }
+
+            #endregion
+        }
+
+        /// <summary>
+        ///   Performs basic smoke test validation of the contained snippet.
+        /// </summary>
+        ///
+        [Test]
+        public async Task ProcessErrorHandlerCancellationTrigger()
```

Comment:
I wonder... since everything that you need to invoke the error handler has a public constructor, is there benefit to having it included in the examples?   The cancellation scenario that I was thinking of is what you're using for the initialization sample.... and I think it fits better there.    I can't come up with a unique thing for error handling that would seem to be confusing enough that a sample would be helpful.

Thoughts?

Reply:
I agree, I spent few hours trying to squeeze something out for demonstrating how factory could be used in the error handling scenario without falling into a particular business case, but just couldn't find anything.

The time was not wasted as I have learned many things along the way. Definitely will be useful in the future.